Add linting implementation

matthew-chirgwin opened the pull request:
- Adds implementation for linting as per #9 
- Vast majority of the changes are the results of the linting being applied across the codebase

Closes: #3 

Signed-off-by: Matthew Chirgwin <chirmatt@uk.ibm.com>

matthew-chirgwin committed:
feat: Add tooling add proposed lint rules

- adds tooling and proposed rules to enable linting

Signed-off-by: Matthew Chirgwin <chirmatt@uk.ibm.com>

matthew-chirgwin committed:
feat: complete linting implementation

- completes linting configuration
- applies configuration to all existing files

Signed-off-by: Matthew Chirgwin <chirmatt@uk.ibm.com>

gmcrobert commented on README.md:
```diff
@@ -19,18 +24,69 @@ Once these prerequisites are met, all required dependencies to build and run the
+
+### Troubleshooting
+
+#### `Signed-off-by` reported as not included in commit, when it is included
```
Wouldn't this be better in a `Known Issues` section than a `Troubleshooting` section?

matthew-chirgwin replied:
I don't mind changing this to Known issues - shall do so in the next commit.

scholzj commented on CODE_OF_CONDUCT.md:
```diff
@@ -1,3 +1,8 @@
+<!--
+ Copyright Strimzi authors.
+ License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+-->
+
```
Can't we just add these files (CODE_OF_CONDUCT.md, GOVERNANCE.md, MAINTAINERS.md etc.) to some ignore list? It will make it a lot easier if we ever need to update them in all XX repositories and can just copy files and not do something special for this repo. (if any ignoring for this would be possible, I would suggest to also use it for all MD files and all files which have nothing to do with code)

matthew-chirgwin replied:
Absolutely, the tool can be configured to ignore any type of file/directories if needed. I will update the config in the next commit to ignore .md files across the board.

gmcrobert commented on utils/headers/README.md:
```diff
@@ -0,0 +1,11 @@
+
+# Headers
+
+This directory contains files with copyright/header statements which will be programmatically added to all source files. These headers are applied at build time to all built output, and at lint time via the `npm run lint:allfiles` script. Files in this folder should be:
```
A little picky again but either drop the word `be` at the end of the sentence as it is included in the bullet points or drop the word off each of the bullet points to read correctly.

matthew-chirgwin replied:
Agreed this reads more nicely, will update in the next commit.

gmcrobert reviewed (changes requested):
Sorry I am late to the party but a couple of general comments about `build.md`.

There is a spelling mistake in the `UI Build` section. The last sentence of that paragraph contains `efficient built for day to day` and I think it should be `efficient build for day to day`.

A question about module aliases. If I am browsing the source code, is there going to be an easy way to map aliased module names in the import statements to source files on the repository without me having to work out what some mangling function generated the alias? I am not expecting you to necessarily do anything but I am interested in knowing how I can very quickly know this as I move around the codebase.

Other than the few typos - the rest looks good to me.

matthew-chirgwin commented:
> There is a spelling mistake in the `UI Build` section. The last sentence of that paragraph contains `efficient built for day to day` and I think it should be `efficient build for day to day`.

I fixed a few other spelling errors in this change, I will fix this one in the next commit as well - good spot!

> A question about module aliases. If I am browsing the source code, is there going to be an easy way to map aliased module names in the import statements to source files on the repository without me having to work out what some mangling function generated the alias? I am not expecting you to necessarily do anything but I am interested in knowing how I can very quickly know this as I move around the codebase.

The alias implementation/documentation as is is the first step of how aliases will be used. In their current form, they allow for convient importing of code. Ie, 

```
import { X } from './View.js';
```

would become

```
import { X } from 'View'
```

As will be covered in https://github.com/strimzi/strimzi-ui/pull/10 when updated, the Aliasing will be used to provide the Swappable view layer capability (so the value of 'View' will be dynamic at build time for instance), as well as developer convience. For instance, via the use of Aliases and [barrel files (a TypeScript codestyle)](https://basarat.gitbook.io/typescript/main-1/barrel), an import such as:

```
import { Foo } from '../path/to/component/Foo/View'
```

becomes:

```
import { Foo } from 'Element';
```

where `Element` is a type of component (again, to be defined in https://github.com/strimzi/strimzi-ui/pull/10 ). In this instance, `Element` would resolve to `../path/to/component/` from the example above, with the barrel files abstracting the directory the component comes from, and where the view.js code is defined.

Similarly, for a custom React hook `useFoo`, it would be:

```
import { useFoo } from 'Hooks';
```

The intent with the Aliasing and structure is to provide convience via abstraction but maintain context of what the thing is your are using is, and where it can be found in the codebase.

matthew-chirgwin committed:
refactor: 1st set of review comments

- do not consider .md files as code - so do not add copyright header to
them
- wording, spelling and grammer changes

Signed-off-by: Matthew Chirgwin <chirmatt@uk.ibm.com>

nictownsend commented on docs/Linting.md:
```diff
@@ -57,6 +52,8 @@ The following must be present in all git commits made to this repository. These
 
 These checks can only be run having made a `git commit`.
 
+_Important note_: If you run into issues with the `All commits must have a signed-off-by entry` rule, [see this troubleshooting item, which covers a workaround for a known bug.](../README.md#signed-off-by-reported-as-not-included-in-commit,-when-it-is-included)
```
Now a known issue :)

matthew-chirgwin replied:
Good spot - thanks :)

nictownsend commented on utils/linting/husky.config.js:
```diff
@@ -0,0 +1,14 @@
+
+// https://github.com/typicode/husky#supported-hooks
+const config = {
+  'pre-commit': 'npm run lint && git add .', // run the lint, and include any auto fixes
```
Is this running risk of adding everything in repo when you might not want to? https://github.com/okonet/lint-staged might be a better answer

matthew-chirgwin replied:
It would indeed add everything to the commit in it's current state. I will look into lint-staged and update as needed

matthew-chirgwin committed:
refactor: introduce lint-staged to linting

- adds lint-staged tool to linting process to only lint staged files
- update docs to cover

Signed-off-by: Matthew Chirgwin <chirmatt@uk.ibm.com>

matthew-chirgwin commented:
@nictownsend thank you for the comments/suggestions - updates have been made in the latest commit to reflect these.

@gmcrobert / @dlabaj did either of you have further thoughts/comments?

dlabaj commented on utils/build/webpack.common.js:
```diff
@@ -23,128 +18,157 @@ const IMAGES_DIR = path.resolve(__dirname, '../../client/Images');
+  },
+  module: {
+    // given the dynamic nature of rules and configuration, none a re provided by default. Use the exported `moduleLoaders` to return loaders with some default configs
+    rules: [],
```
Do we need these extra commas at the end of all these json objects in webpack files?

dlabaj reviewed (commented):
Looks good ... only question was wondering if we can remove those trailing , in the webpack files.

matthew-chirgwin commented:
> Looks good ... only question was wondering if we can remove those trailing , in the webpack files.

Thanks for the review @dlabaj - the commas are a part of prettier's default settings. As a part of https://github.com/strimzi/strimzi-ui/issues/4 we can discuss what settings we want, and implement those. The prupose of this PR was just to add the tooling and base implementation to support.

dlabaj commented:
> > Looks good ... only question was wondering if we can remove those trailing , in the webpack files.
> 
> Thanks for the review @dlabaj - the commas are a part of prettier's default settings. As a part of #4 we can discuss what settings we want, and implement those. The prupose of this PR was just to add the tooling and base implementation to support.

Thanks Matthew.  Approved.

nictownsend reviewed (approved):
LGTM, nice config for lint staged 👍
